server/src/models/pmcmean.rs

```diff
@@ -13,46 +13,131 @@
+    /// otherwise `false`.
+    fn fit_value(&mut self, value: f32) -> bool {
+        let next_min_value = f32::min(self.min_value, value);
+        let next_max_value = f32::max(self.max_value, value);
+        let next_sum_of_values = self.sum_of_values + value as f64;
+        let next_length = self.length + 1;
+        let average = (next_sum_of_values / next_length as f64) as f32;
+        if self.is_value_within_error_bound(next_min_value, average)
+            && self.is_value_within_error_bound(next_max_value, average)
+        {
+            self.min_value = next_min_value;
+            self.max_value = next_max_value;
+            self.sum_of_values = next_sum_of_values;
+            self.length = next_length;
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Return the current model. For a model of type PMC-Mean, its coefficient
+    /// is the average value of the time series segment the model represents.
+    fn get_model(&self) -> f32 {
+        (self.sum_of_values / self.length as f64) as f32
+    }
+
+    /// Determine if `approximate_value` is within the relative
+    /// `self.error_bound` of `real_value`.
+    fn is_value_within_error_bound(&self, real_value: f32, approximate_value: f32) -> bool {
+        // Needed because result becomes NAN and approximate_value is rejected
+        // if approximate_value and real_value are zero, and because NAN != NAN.
+        if equal_or_nan(real_value, approximate_value) {
+            true
+        } else {
+            let difference = real_value - approximate_value;
+            let result = f32::abs(difference / real_value);
+            (result * 100.0) <= self.error_bound
+        }
+    }
+}
+
+/// Compute the minimum and maximum value for a time series segment whose values
+/// are represented by a model of type PMC-Mean.
 pub fn min_max(
-    gid: i32,
-    start_time: i64,
-    end_time: i64,
-    sampling_interval: i32,
+    _gid: i32,
+    _start_time: i64,
+    _end_time: i64,
+    _sampling_interval: i32,
     model: &[u8],
-    gaps: &[u8],
+    _gaps: &[u8],
 ) -> f32 {
-    decode(model)
+    decode_model(model)
 }
 
+/// Compute the sum of the values for a time series segment whose values are
+/// represented by a model of type PMC-Mean.
 pub fn sum(
-    gid: i32,
+    _gid: i32,
     start_time: i64,
     end_time: i64,
     sampling_interval: i32,
     model: &[u8],
-    gaps: &[u8],
+    _gaps: &[u8],
 ) -> f32 {
     let length = ((end_time - start_time) / sampling_interval as i64) + 1;
-    length as f32 * decode(model)
+    length as f32 * decode_model(model)
 }
 
+/// Reconstruct the data points for a time series segment whose values are
+///  represented by a model of type PMC-Mean.
 pub fn grid(
     gid: i32,
     start_time: i64,
     end_time: i64,
     sampling_interval: i32,
     model: &[u8],
-    gaps: &[u8],
+    _gaps: &[u8],
     tids: &mut Int32Builder,
     timestamps: &mut TimestampMillisecondBuilder,
```

Comment:
Since the MQTT simulator generates timestamps with microsecond precision, we currently use a builder with TimestampMicrosecondType in the storage engine.

To avoid further confusion we should probably define type alias' in main that specify the type for single values, the arrays and the builders.

Reply:
I have added the type aliases to a new module named [types.rs](https://github.com/ModelarData/MiniModelarDB/blob/e888c7f608b1608d7f0c50425477feb69cf323d9/server/src/types.rs) so we have a central place to define types used throughout the system. For the moment the types are defined based on the requirements of [pmcmean.rs](https://github.com/ModelarData/MiniModelarDB/blob/e888c7f608b1608d7f0c50425477feb69cf323d9/server/src/models/pmcmean.rs) as changing its types also requires changes to [tables.rs](https://github.com/ModelarData/MiniModelarDB/blob/e888c7f608b1608d7f0c50425477feb69cf323d9/server/src/tables.rs).